### PR TITLE
feat(spdx): Add API for SPDX Document tab

### PR DESCRIPTION
**Pull request by tuannn2**
[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

- Get SPDX Document Information
- Update SPDX Document Information

#### 1. Get SPDX Document information
- Run command line: 
```
curl 'https://sw360.org/api/releases/3765276512' -i -X GET \
    -H 'Authorization: Bearer eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOlsic3czNjAtUkVTVC1BUEkiXSwidXNlcl9uYW1lIjoiYWRtaW5Ac3czNjAub3JnIiwic2NvcGUiOlsiYWxsIl0sImV4cCI6MTcxMjkxOTg2NywiYXV0aG9yaXRpZXMiOlsiUkVBRCIsIldSSVRFIl0sImp0aSI6IlFNd3UxMFVWdUUwdFd3MkV5VnRsQm1OdGRzdyIsImNsaWVudF9pZCI6InRydXN0ZWQtc3czNjAtY2xpZW50In0.Xp3TdSvb2oQ77zqg8RLaz4F47vFEX0lPL4ijmKRmL9udAUWvV0GMWYeKVwYAQSDlPYPwqGxSk66wRp0R7hrpemKEAfmfA2X5V6-xwUaPsnV9nklr0wm_PGHxQDQIwI9yHDXwkRw9sYtFH67oQt32KRAJOeQnAkzT2Hgy3Ok45_XRmhQM9xcISdEOHN3svf6719pJbRJ-YRtLiRqhGaJGqbiCB6Iixu55zx0buqZc0PMPP-lwHQksiMGujXj1yQGeSt5FChuk87cxSqSLkkD56RVYUDeTa5g6kUsFdWHOHC6vpzT0TK5ihP_y2JhfvNKWy6nLgNxnqaQavvaOeoXZoA' \
    -H 'Accept: application/hal+json'
```
#### 2. Update SPDX Document information
- Run command line: 
```
curl 'https://sw360.org/api/releases/12121212/spdx' -i -X PATCH \
    -H 'Content-Type: application/hal+json' \
    -H 'Authorization: BearereyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOlsic3czNjAtUkVTVC1BUEkiXSwidXNlcl9uYW1lIjoiYWRtaW5Ac3czNjAub3JnIiwic2NvcGUiOlsiYWxsIl0sImV4cCI6MTcxMjkxOTg2NywiYXV0aG9yaXRpZXMiOlsiUkVBRCIsIldSSVRFIl0sImp0aSI6IjU3V0FjbWdBamFsTnNlNktXTE9qNzd4MzJENCIsImNsaWVudF9pZCI6InRydXN0ZWQtc3czNjAtY2xpZW50In0.bqnOCLGdOiYGj8liZazzGJw2ce1ahXrnv7nJxHM47J0E3MoQsHea7DAOhPYLNubZMFpjq8ynHwss8ZsxtoKFdJ8V_7zy5M4mpECwybOSqmx94qksl3vDjQwOOpGTWLshezZ1k8w1w-LeA9CecNSe9t_0K-kdOoGMDMGYfDuVQLOv67QGtuxYI--ZlH_xh-wc6AzLHFuh6rTkoi14N50OI5jBI8PK-bajdcce1m0KCPplgVwMWX5cu9hpvLcDJOq9EYf3ezWQzlcMHNDRR-_CNkKbs1XX4Jsri-KZEssSB8dQ9C2LZA5SK1mQartmXcY3rINPxVh5XmLDwufzrjcymQ' \
    -H 'Accept: application/hal+json' \
    -d '{
  "packageInformation" : {
    "type" : "packageInformation",
    "name" : "11",
    "SPDXID" : "SPDXRef-Package-11",
    "versionInfo" : "11",
    "packageFileName" : "11",
    "supplier" : "Organization: 11",
    "originator" : "Organization: 11",
    "downloadLocation" : "11",
    "filesAnalyzed" : true,
    "packageVerificationCode" : {
      "excludedFiles" : [ "11" ],
      "value" : "11"
    },
    "checksums" : [ {
      "algorithm" : "11",
      "checksumValue" : "1111",
      "index" : 0
    } ],
    "homepage" : "11",
    "sourceInfo" : "11",
    "licenseConcluded" : "11",
    "licenseInfoFromFiles" : [ "11" ],
    "licenseDeclared" : "11",
    "licenseComments" : "11",
    "copyrightText" : "11",
    "summary" : "11",
    "description" : "11",
    "packageComment" : "11",
    "externalRefs" : [ {
      "referenceCategory" : "SECURITY",
      "referenceLocator" : "11",
      "referenceType" : "cpe22Type",
      "comment" : "11",
      "index" : 0
    } ],
    "attributionText" : [ "11" ],
    "annotations" : [ ],
    "primaryPackagePurpose" : "11",
    "releaseDate" : "2023-11-10T07:30:39Z",
    "builtDate" : "2023-11-09T07:29:43Z",
    "validUntilDate" : "2023-11-22T07:30:47Z",
    "relationships" : [ {
      "spdxElementId" : "11",
      "relationshipType" : "11",
      "relatedSpdxElement" : "11",
      "relationshipComment" : "11",
      "index" : 0
    } ],
    "moderators" : [ ]
  },
  "spdxDocument" : {
    "type" : "SPDXDocument",
    "spdxFileInfoIds" : [ ],
    "snippets" : [ {
      "SPDXID" : "SPDXRef-Snippet-11",
      "snippetFromFile" : "SPDXRef-11",
      "snippetRanges" : [ {
        "rangeType" : "11",
        "startPointer" : "11",
        "endPointer" : "11",
        "reference" : "11",
        "index" : 0
      } ],
      "licenseConcluded" : "11",
      "licenseInfoInSnippets" : [ "11" ],
      "licenseComments" : "11",
      "copyrightText" : "11",
      "comment" : "11",
      "name" : "11",
      "snippetAttributionText" : "11",
      "index" : 0
    } ],
    "relationships" : [ {
      "spdxElementId" : "11",
      "relationshipType" : "11",
      "relatedSpdxElement" : "11",
      "relationshipComment" : "11",
      "index" : 0
    } ],
    "annotations" : [ {
      "annotator" : "Organization: 11",
      "annotationDate" : "2023-11-14T07:31:11Z",
      "annotationType" : "11",
      "spdxIdRef" : "11",
      "annotationComment" : "11",
      "index" : 0
    } ],
    "otherLicensingInformationDetecteds" : [ {
      "licenseId" : "LicenseRef-11",
      "extractedText" : "11",
      "licenseName" : "11",
      "licenseCrossRefs" : [ "11" ],
      "licenseComment" : "11",
      "index" : 0
    } ],
    "documentState" : {
      "isOriginalDocument" : true,
      "moderationState" : null,
      "setModerationState" : false,
      "setIsOriginalDocument" : true
    },
    "moderators" : [ ]
  },
  "documentCreationInformation" : {
    "type" : "documentCreationInformation",
    "spdxVersion" : "SPDX-3333",
    "dataLicense" : "11",
    "SPDXID" : "SPDXRef-1111",
    "name" : "11",
    "documentNamespace" : "11",
    "externalDocumentRefs" : [ {
      "externalDocumentId" : "11",
      "checksum" : {
        "algorithm" : "11",
        "checksumValue" : "11",
        "index" : 0
      },
      "spdxDocument" : "11",
      "index" : 0
    } ],
    "licenseListVersion" : "11",
    "creator" : [ {
      "type" : "Person",
      "value" : "Test Admin (admin@sw360.org)",
      "index" : 0
    } ],
    "created" : "2023-11-27T07:25:40Z",
    "creatorComment" : "11",
    "documentComment" : "11",
    "createdBy" : "admin@sw360.org",
    "moderators" : [ ]
  }
}'
```

### Checklist
Must:
- [ ] All related issues are referenced in commit messages and in PR

### CODE REVIEW

**Comment by GMishx** [comment]
@tuannn2 please rebase the branch for resolving merge conflict.

@rudra-superrr @afsahsyeda please help with review/test.

**Comment by hoangnt2** [comment]
@KoukiHama, I am checking it

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/JacksonCustomizations.java [hunk 1]
```diff
@@ -82,6 +95,19 @@ public Sw360Module() {
...
+            setMixInAnnotation(CheckSum.class, Sw360Module.CheckSumMixin.class);
+            setMixInAnnotation(ExternalDocumentReferences.class, Sw360Module.ExternalDocumentReferencesMixin.class);
+            setMixInAnnotation(Creator.class, Sw360Module.CreatorMixin.class);
+            setMixInAnnotation(PackageVerificationCode.class, Sw360Module.PackageVerificationCodeMixin.class);
+            setMixInAnnotation(ExternalReference.class, Sw360Module.ExternalReferenceMixin.class);
+            setMixInAnnotation(SnippetRange.class, Sw360Module.SnippetRangeMixin.class);
+            setMixInAnnotation(Annotations.class, Sw360Module.AnnotationsMixin.class);
+            setMixInAnnotation(RelationshipsBetweenSPDXElements.class, Sw360Module.RelationshipsBetweenSPDXElementsMixin.class);
+            setMixInAnnotation(SnippetInformation.class, Sw360Module.SnippetInformationMixin.class);
+            setMixInAnnotation(OtherLicensingInformationDetected.class, Sw360Module.OtherLicensingInformationDetectedMixin.class);
```
Comment (GMishx) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/JacksonCustomizations.java:110]: Please add the same mixins for `SpringDocUtils.getConfig().replaceWithClass()` for OpenAPI docs.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/RestControllerHelper.java [hunk 2]
```diff
@@ -1655,4 +1700,357 @@ public ClearingRequest updateCRSize(ClearingRequest clearingRequest, Project pro
         }
         return clearingRequestService.getClearingRequestById(clearingRequest.getId(), sw360User);
     }
+
+    public SPDXDocument generateSpdxDocument() {
```
Comment (GMishx) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/RestControllerHelper.java:1704]: These generators are defined as `public static` in `SpdxDocumentDatabaseHandler`. Can't we reuse them?

Or move them to a util class if backend is not a dependency of Rest module.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java [hunk 3]
```diff
@@ -584,6 +610,70 @@ public ResponseEntity<EntityModel<Release>> createRelease(
         return ResponseEntity.created(location).body(halResource);
     }
 
+    @PreAuthorize("hasAuthority('WRITE')")
+    @PatchMapping(value = RELEASES_URL + "/{id}/spdx")
+    public ResponseEntity<?> updateSPDX( @RequestBody Map<String, Object> reqBodyMap, @PathVariable("id") String releaseId) throws TException {
```
Comment (GMishx) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java:615]: Please document the endpoint with `@Operation` annotation for OpenAPI docs.

Comment (GMishx) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java:615]: And the parameters as well with `@Parameter` annotation.

**Review by GMishx — CHANGES_REQUESTED** [review]
Some suggestions.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/RestControllerHelper.java [hunk 4]
```diff
@@ -1655,4 +1700,357 @@ public ClearingRequest updateCRSize(ClearingRequest clearingRequest, Project pro
...
+        String spdxId = addSPDXDocument(release, user);
+        if (CommonUtils.isNotNullEmptyOrWhitespace(spdxId)) {
+            throw new HttpMessageNotReadableException("Add SPDXDocument Failed!");
+        }
+        addDocumentCreationInformation(spdxId, release.getModerators(), user);
+        addPackageInformation(spdxId, release.getModerators(), user);
+        return spdxId;
+    }
+
+    public boolean checkIndexCreator(Set<Creator> creators) {
```
Comment (GMishx) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/RestControllerHelper.java:1905]: Adding all these validation functions to `RestControllerHelper` does not seem right to me. But I cannot think of any other better place to put them ATM.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/JacksonCustomizations.java (hunk 1 above)
Comment (hoangnt2) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/JacksonCustomizations.java:110]: Added them for SpringDocUtils

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/RestControllerHelper.java (hunk 2 above)
Comment (hoangnt2) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/core/RestControllerHelper.java]: I moved them to SW360Utils

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java (hunk 3 above)
Comment (hoangnt2) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java]: Added `@Operation` and `@Parameter`

**Comment by hoangnt2** [comment]
@GMishx Thanks for your comment. I fixed them. Please help me to review them

**Review by GMishx — DISMISSED** [review]
Changes looks good.

**Comment by GMishx** [comment]
Dear @hoangnt2 / @tuannn2 , I found a bug in the Cloudant code while testing this branch. It is a small change where conversion of Java object to Cloudant document sets `id` and `rev` of a document as an empty string instead of `null` causing a document creation error. Can you please apply the patch to the same PR to expedite the fix?

```diff
diff --git a/libraries/datahandler/src/main/java/org/eclipse/sw360/datahandler/cloudantclient/DatabaseConnectorCloudant.java b/libraries/datahandler/src/main/java/org/eclipse/sw360/datahandler/cloudantclient/DatabaseConnectorCloudant.java
index 78a124bc8..7cff9390d 100644
--- a/libraries/datahandler/src/main/java/org/eclipse/sw360/datahandler/cloudantclient/DatabaseConnectorCloudant.java
+++ b/libraries/datahandler/src/main/java/org/eclipse/sw360/datahandler/cloudantclient/DatabaseConnectorCloudant.java
@@ -666,23 +666,33 @@ public class DatabaseConnectorCloudant {
         Type t = new TypeToken<Map<String, Object>>() {}.getType();
         Map<String, Object> map = gson.fromJson(gson.toJson(document), t);
         if (map.containsKey("id")) {
-            doc.setId((String) map.get("id"));
+            if (!((String) map.get("id")).isEmpty()) {
+                doc.setId((String) map.get("id"));
+            }
             map.remove("id");
         }
         if (map.containsKey("_id")) {
-            doc.setId((String) map.get("_id"));
+            if (!((String) map.get("_id")).isEmpty()) {
+                doc.setId((String) map.get("_id"));
+            }
             map.remove("_id");
         }
         if (map.containsKey("rev")) {
-            doc.setRev((String) map.get("rev"));
+            if (!((String) map.get("rev")).isEmpty()) {
+                doc.setRev((String) map.get("rev"));
+            }
             map.remove("rev");
         }
         if (map.containsKey("revision")) {
-            doc.setRev((String) map.get("revision"));
+            if (!((String) map.get("revision")).isEmpty()) {
+                doc.setRev((String) map.get("revision"));
+            }
             map.remove("revision");
         }
         if (map.containsKey("_rev")) {
-            doc.setRev((String) map.get("_rev"));
+            if (!((String) map.get("_rev")).isEmpty()) {
+                doc.setRev((String) map.get("_rev"));
+            }
             map.remove("_rev");
         }
         doc.setProperties(map);
```

**Comment by hoangnt2** [comment]
@GMishx Sure, I will do that

**Comment by hoangnt2** [comment]
@GMishx I pushed a new commit to apply your patch

**Review by GMishx — APPROVED** [review]
Tested, working as expected.

**Pull request merged**